dhall/src/Dhall/Parser/Expression.hs

```diff
@@ -963,6 +963,29 @@ parsers embedded = Parsers {..}
 
             nonEmptyListLiteral <|> emptyListLiteral) <?> "literal"
 
+-- | Parse a single comment
+comment :: Parser Comment
+comment = LineComment <$> lineComment <|> BlockComment <$> blockComment <?> "comment"
+
+-- | Parse a multi comment (which contains at least one comment)
+multiComment :: Parser MultiComment
+multiComment = setCommentControl CommentIsNeeded $ -- Strict whitespace
+    MultiComment <$> Combinators.NonEmpty.sepEndBy1 comment whitespace
+
+-- | Parse a multi comment or whitespace
+commentOrWhitespace :: Parser (Maybe MultiComment)
+commentOrWhitespace = setCommentControl CommentIsNeeded $ -- Strict whitespace
+    do whitespace; optional multiComment
+
+-- | Parse a multi comment or non-empty whitespace
+commentOrNonEmptyWhitespace :: Parser (Maybe MultiComment)
+commentOrNonEmptyWhitespace = setCommentControl CommentIsNeeded $ do -- Strict whitespace
```

Comment:
I suggest moving the `--Strict whitespace` comment to the documentation for the `CommentIsNeeded` constructor

```suggestion
-- | Parse a multi comment (which contains at least one comment)
multiComment :: Parser MultiComment
multiComment = setCommentControl CommentIsNeeded $
    MultiComment <$> Combinators.NonEmpty.sepEndBy1 comment whitespace

-- | Parse a multi comment or whitespace
commentOrWhitespace :: Parser (Maybe MultiComment)
commentOrWhitespace = setCommentControl CommentIsNeeded $
    do whitespace; optional multiComment

-- | Parse a multi comment or non-empty whitespace
commentOrNonEmptyWhitespace :: Parser (Maybe MultiComment)
commentOrNonEmptyWhitespace = setCommentControl CommentIsNeeded $ do
```